utils/near-cache/src/lib.rs

```diff
@@ -10,37 +10,46 @@ pub struct SyncLruCache<K, V> {
 
 impl<K, V> SyncLruCache<K, V>
 where
-    K: Hash + Eq + PartialEq + Clone,
+    K: Hash + Eq,
     V: Clone,
 {
-    /// Creates new cache with `capacity`.
-    pub fn new(capacity: usize) -> Self {
-        Self { inner: Mutex::new(LruCache::<K, V>::new(capacity)) }
+    /// Creats a new `LRU` cache that holds at most `cap` items.
+    pub fn new(cap: usize) -> Self {
+        Self { inner: Mutex::new(LruCache::<K, V>::new(cap)) }
     }
 
-    /// Gets value for `key`
-    pub fn get_or_insert(&self, key: K, f: impl FnOnce(&K) -> V) -> V
-    where
-        V: Clone,
-    {
-        if let Some(result) = self.get(&key) {
-            return result;
-        }
-        let val = f(&key);
-        let val_clone = val.clone();
-        self.inner.lock().unwrap().put(key, val_clone);
-        val
+    /// Return the value of they key in the cache otherwise computes the value and inserts it into
+    /// the cache. If the key is already in the cache, they gets gets moved to the head of
+    /// the LRU list.
+    pub fn get_or_put(&self, key: K, f: impl FnOnce(&K) -> V) -> V {
+        self.get(&key).unwrap_or_else(|| {
+            let val = f(&key);
+            self.inner.lock().unwrap().put(key, val.clone());
```

Comment:
Note that this re-introduces a minor perf bug where we clone the value in the scope of the lock. That's a rule-of-thumb: everywhere you do `.lock()`, take a second look at the code and see if you can minimize the extent at which the lock is held.